In [2]:
import seaborn as sns
sns.set()

In [3]:
%load_ext sql
%sql sqlite:///small_data/customers.sqlite

'Connected: @small_data/customers.sqlite'

get a list of all the tables in this database

In [37]:
%%sql

SELECT * FROM sqlite_master WHERE type='table';

 * sqlite:///small_data/customers.sqlite
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'adclicks', 'adclicks', 4, 'CREATE TABLE "adclicks" (\n"time" TEXT,\n  "adID" INTEGER,\n  "clicked" INTEGER,\n  "webpageViewed" TEXT,\n  "userID" INTEGER\n)'),
 ('table', 'customers', 'customers', 19, 'CREATE TABLE "customers" (\n\t"userID"\tINTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,\n\t"name"\tTEXT,\n\t"zipCode"\tTEXT,\n\t"age"\tINTEGER\n)'),
 ('table', 'zip_state', 'zip_state', 2, 'CREATE TABLE "zip_state" (\n"zip" TEXT,\n  "city" TEXT,\n  "county" TEXT,\n  "state" TEXT\n)')]

get the schema for the table directly, to look at the types of each column

In [39]:
%%sql

PRAGMA table_info(customers)

 * sqlite:///small_data/customers.sqlite
Done.


[(0, 'userID', 'INTEGER', 1, None, 1),
 (1, 'name', 'TEXT', 0, None, 0),
 (2, 'zipCode', 'TEXT', 0, None, 0),
 (3, 'age', 'INTEGER', 0, None, 0)]

In [40]:
%%sql

SELECT * FROM customers LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[(1, 'Katie Baker', '13083', 54),
 (2, 'Mary Clarkson', '07852', 20),
 (3, 'Michelle Reynolds', '06403', 26),
 (5, "Valerie O'Brian", '11386', None),
 (7, 'Katherine Diamond', '08605', 26),
 (8, 'Imran Brown', '06794', 40),
 (10, 'Francis Baker', None, 72),
 (11, 'Rob Smith', '07442', 5),
 (12, 'Clare Johnson', '06410', 53),
 (13, 'Gabriel Thomasson', None, 19)]

In [41]:
result = _
print(type(result))
print(result)

<class 'sql.run.ResultSet'>
+--------+-------------------+---------+------+
| userID |        name       | zipCode | age  |
+--------+-------------------+---------+------+
|   1    |    Katie Baker    |  13083  |  54  |
|   2    |   Mary Clarkson   |  07852  |  20  |
|   3    | Michelle Reynolds |  06403  |  26  |
|   5    |  Valerie O'Brian  |  11386  | None |
|   7    | Katherine Diamond |  08605  |  26  |
|   8    |    Imran Brown    |  06794  |  40  |
|   10   |   Francis Baker   |   None  |  72  |
|   11   |     Rob Smith     |  07442  |  5   |
|   12   |   Clare Johnson   |  06410  |  53  |
|   13   | Gabriel Thomasson |   None  |  19  |
+--------+-------------------+---------+------+


In [42]:
print(result[0].name)
print(result[0][1])

Katie Baker
Katie Baker


In [43]:
result = %sql SELECT * FROM customers LIMIT 20;
print(result)

 * sqlite:///small_data/customers.sqlite
Done.
+--------+-------------------+---------+------+
| userID |        name       | zipCode | age  |
+--------+-------------------+---------+------+
|   1    |    Katie Baker    |  13083  |  54  |
|   2    |   Mary Clarkson   |  07852  |  20  |
|   3    | Michelle Reynolds |  06403  |  26  |
|   5    |  Valerie O'Brian  |  11386  | None |
|   7    | Katherine Diamond |  08605  |  26  |
|   8    |    Imran Brown    |  06794  |  40  |
|   10   |   Francis Baker   |   None  |  72  |
|   11   |     Rob Smith     |  07442  |  5   |
|   12   |   Clare Johnson   |  06410  |  53  |
|   13   | Gabriel Thomasson |   None  |  19  |
|   14   |    Louis Zipes    |  06267  |  56  |
|   15   |   Mary Lovelace   |  14005  |  39  |
|   16   |  Francis Polanski |  13413  |  50  |
|   17   |   William Cooper  |  14005  |  59  |
|   19   |    Charles Carr   |  06794  |  32  |
|   20   |    Violet Green   |  06794  |  20  |
|   21   |  David Gunderson  |  13083  | 

In [6]:
%%sql 
SELECT name, age FROM customers LIMIT 5;

 * sqlite:///small_data/customers.sqlite
Done.


[('Katie Baker', 54),
 ('Mary Clarkson', 20),
 ('Michelle Reynolds', 26),
 ("Valerie O'Brian", None),
 ('Katherine Diamond', 26)]

In [38]:
%%sql

SELECT * FROM adclicks LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[('2019-05-04 15:00:39', 9, 0, 'ecommerce', 64),
 ('2019-12-13 01:29:48', 8, 0, 'news', 48),
 ('2019-12-05 04:19:10', 10, 0, 'social', 22),
 ('2019-10-13 14:08:22', 6, 0, 'ecommerce', 30),
 ('2019-08-14 15:23:00', 2, 0, 'news', 97),
 ('2019-02-23 14:58:16', 8, 0, 'news', 44),
 ('2019-02-11 01:00:18', 8, 0, 'social', 84),
 ('2019-03-30 12:04:26', 6, 0, 'social', 129),
 ('2019-09-16 16:11:23', 8, 0, 'social', 35),
 ('2019-01-21 21:02:31', 10, 0, 'social', 7)]

find the number of customers in the customers table.

In [7]:
%%sql 
SELECT COUNT(*) FROM customers;  

 * sqlite:///small_data/customers.sqlite
Done.


[(113,)]

In [8]:
%%sql 
SELECT SUM(clicked)*1.0/COUNT(clicked) FROM adclicks;

 * sqlite:///small_data/customers.sqlite
Done.


[(0.11506434519303559,)]

So the basic click-through rate is about 11.5%

Counts distinct non-null(!) entries

In [9]:
%%sql 
SELECT COUNT(DISTINCT zipcode) AS number_of_zip_codes 
FROM customers;

 * sqlite:///small_data/customers.sqlite
Done.


[(20,)]

In [10]:
%%sql
SELECT * 
FROM customers 
WHERE age > 50 LIMIT 10;


 * sqlite:///small_data/customers.sqlite
Done.


[(1, 'Katie Baker', '13083', 54),
 (10, 'Francis Baker', None, 72),
 (12, 'Clare Johnson', '06410', 53),
 (14, 'Louis Zipes', '06267', 56),
 (17, 'William Cooper', '14005', 59),
 (21, 'David Gunderson', '13083', 55),
 (29, 'Francis Reynolds', '08605', 53),
 (34, 'Valerie Clarkson', '07416', 52),
 (51, 'Ann Wallace', '10065', 56),
 (55, 'John Brown', '13413', 78)]

In [11]:
%%sql 
SELECT * 
FROM customers 
WHERE age > 50 AND zipcode = '11715';

 * sqlite:///small_data/customers.sqlite
Done.


[(77, 'Clare Lovelace', '11715', 55), (116, 'Kate Vegas', '11715', 56)]

In [12]:
%%sql
SELECT SUM(clicked)*1.0/COUNT(clicked) AS e_commerce_rate 
FROM adclicks 
WHERE webpageviewed = 'ecommerce';

 * sqlite:///small_data/customers.sqlite
Done.


[(0.0784313725490196,)]

So when customers see advertisements on E-commerce websites the ad click-through rate is lower than the overall average, only about 7.8%.

Get ZIP codes that start with '06'

In [13]:
%%sql 
SELECT * 
FROM customers 
WHERE zipcode LIKE '06%' LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[(3, 'Michelle Reynolds', '06403', 26),
 (8, 'Imran Brown', '06794', 40),
 (12, 'Clare Johnson', '06410', 53),
 (14, 'Louis Zipes', '06267', 56),
 (19, 'Charles Carr', '06794', 32),
 (20, 'Violet Green', '06794', 20),
 (32, 'Jonathan Barker', '06794', 33),
 (36, 'Chuck Finley', '06267', 16),
 (37, 'Rob Frye', '06535', 40),
 (38, 'Mary Reynolds', '06794', 23)]

In [14]:
%%sql 
SELECT COUNT(*) 
FROM customers 
WHERE zipcode LIKE '06%' AND age > 50;

 * sqlite:///small_data/customers.sqlite
Done.


[(4,)]

We were told by the marketing department that there is information for 10 advertisements in the given data. Use DISTINCT to find the missing adid in the adclicks table. Using ORDER BY in the query will make it easier to identify which value is missing.

In [15]:
%%sql 
SELECT DISTINCT adid 
FROM adclicks 
ORDER BY adid;    

 * sqlite:///small_data/customers.sqlite
Done.


[(1,), (2,), (4,), (5,), (6,), (7,), (8,), (9,), (10,)]

In [16]:
%%sql 
SELECT * 
FROM customers  
ORDER BY zipcode DESC, age DESC;

 * sqlite:///small_data/customers.sqlite
Done.


[(17, 'William Cooper', '14005', 59),
 (15, 'Mary Lovelace', '14005', 39),
 (80, 'Imran Polanski', '14005', 28),
 (119, 'Michelle Frye', '14005', 22),
 (25, 'Phillip Goldberg', '13654', 28),
 (55, 'John Brown', '13413', 78),
 (122, 'Gary Lovelace', '13413', 53),
 (16, 'Francis Polanski', '13413', 50),
 (94, 'Mary Gunderson', '13413', 42),
 (33, 'William Brand', '13413', 41),
 (83, "William O'Brian", '13413', 41),
 (109, 'Beth Brown', '13413', 41),
 (31, "Katie O'Brian", '13413', 31),
 (72, 'Julia', '13413', 31),
 (75, 'Ann Smith', '13413', 22),
 (62, 'Violet Gunderson', '13413', 20),
 (21, 'David Gunderson', '13083', 55),
 (1, 'Katie Baker', '13083', 54),
 (59, 'Ann Gunderson', '13083', 39),
 (112, 'Katie Frye', '13083', 39),
 (49, 'Katie Davidson', '13083', 33),
 (66, 'Irina Brown', '13083', 18),
 (130, 'George Smith', '13083', 17),
 (46, 'Peter Kirkpatrick', '11731', 49),
 (48, 'Gary Reynolds', '11731', 48),
 (50, 'Katie Gunderson', '11731', 36),
 (102, 'Valerie Cooper', '11731', 29),
 (56, 'Beth Vegas', '11731', 26),
 (23, 'Marie Frye', '11731', 25),
 (30, 'Charles Baker', '11731', -1),
 (116, 'Kate Vegas', '11715', 56),
 (77, 'Clare Lovelace', '11715', 55),
 (43, 'Charles Goldberg', '11715', 47),
 (40, 'Charles Reeves', '11715', 42),
 (28, 'Kate Cooper', '11715', 40),
 (47, 'Beth Frye', '11715', 34),
 (22, 'Beth Reynolds', '11715', 29),
 (97, 'William Gunderson', '11715', 29),
 (42, 'Charles Wallace', '11715', 25),
 (100, 'Gary Goldberg', '11386', 82),
 (95, 'Marie Clarkson', '11386', 47),
 (90, 'Irina Smith', '11386', 44),
 (44, 'George Baker', '11386', 42),
 (84, 'Ann Cooper', '11386', 42),
 (88, 'Michelle Reeves', '11386', 41),
 (63, 'William Polanski', '11386', 16),
 (5, "Valerie O'Brian", '11386', None),
 (111, 'Gary Davidson', '10065', 58),
 (51, 'Ann Wallace', '10065', 56),
 (67, 'Ann Clarkson', '10065', 49),
 (99, 'Charles Lovelace', '10065', 46),
 (103, 'Katie Vegas', '10065', 39),
 (113, 'David Brown', '10065', 32),
 (61, 'Irina Baker', '10065', 31),
 (115, 'David Armstrong', '10065', 16),
 (57, 'Phillip Lovelace', '08810', 56),
 (91, "Rob O'Brian", '08810', 43),
 (118, 'Katie Reynolds', '08810', 21),
 (27, 'Michelle Baker', '08810', 1),
 (70, 'Michelle Smith', '08605', 57),
 (29, 'Francis Reynolds', '08605', 53),
 (87, 'Marie Gunderson', '08605', 39),
 (123, 'Marie Rosenthal', '08605', 27),
 (7, 'Katherine Diamond', '08605', 26),
 (126, 'Prudence Cooper', '08203', 39),
 (53, 'Marie Johansen', '08203', 23),
 (127, 'Georgina Freeman', '08203', 21),
 (96, 'Javier Luis Bolivar de Vasquez Ibarra', '08203', 16),
 (73, 'Johnny Vegas', '08203', None),
 (69, 'David Baker', '07852', 54),
 (78, None, '07852', 26),
 (2, 'Mary Clarkson', '07852', 20),
 (98, 'Mary Smith', '07442', 56),
 (81, 'David Wallace', '07442', 37),
 (92, 'Vladimir Stravinski', '07442', 33),
 (11, 'Rob Smith', '07442', 5),
 (114, 'Belinda Blumenthal', '07416', 57),
 (34, 'Valerie Clarkson', '07416', 52),
 (110, 'Marie Lovelace', '07416', 29),
 (71, 'Imran Gunderson', '07416', 25),
 (58, 'Elizabeth Anderson', '07416', 24),
 (129, 'Katie Brand', '06794', 58),
 (8, 'Imran Brown', '06794', 40),
 (32, 'Jonathan Barker', '06794', 33),
 (19, 'Charles Carr', '06794', 32),
 (38, 'Mary Reynolds', '06794', 23),
 (20, 'Violet Green', '06794', 20),
 (60, 'Marie Brand', '06794', 19),
 (89, 'Valerie Wallace', '06794', 17),
 (104, 'Katie Clarkson', '06535', 57),
 (64, 'Damien Cruise', '06535', 49),
 (82, 'Josephine Smith', '06535', 46),
 (37, 'Rob Frye', '06535', 40),
 (54, 'Hugh Jackson', '06535', 33),
 (65, 'Phillip Wallace', '06535', None),
 (12, 'Clare Johnson', '06410', 53),
 (41, 'Irina Johansen', '06410', 21),
 (39, 'Julia Powers', '06403', 37),
 (52, 'Bobby Brand', '06403', 31),
 (3, 'Michelle Reynolds', '06403', 26),
 (93, 'George Frye', '06403', 25),
 (14, 'Louis Zipes', '06267', 56),
 (79, 'Mary Brand', '06267', 42),
 (74, 'Mary Clarkson', '06267', 31),
 (36, 'Chuck Finley', '06267', 16),
 (121, 'Irina Goldberg', '06245', 33),
 (128, 'Yolanda Richardson', '0

What ZIP code(s) has (have) the largest number of customers in it/them? (Hint: Use ORDER BY to order the group counts in descending order. It might help to alias the group counts and use that alias in the ORDER BY clause.)

In [17]:
%%sql 
SELECT zipcode, COUNT(*) AS counts 
FROM customers 
GROUP BY zipcode ORDER BY counts DESC LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[('13413', 11),
 ('11715', 9),
 ('11386', 8),
 ('10065', 8),
 ('06794', 8),
 ('13083', 7),
 ('11731', 7),
 ('06535', 6),
 (None, 6),
 ('08605', 5)]

what ZIP codes have 10 or more customers in them

In [18]:
%%sql 
SELECT zipcode, COUNT(*) AS number 
FROM customers 
GROUP BY zipcode 
HAVING number >= 10;

 * sqlite:///small_data/customers.sqlite
Done.


[('13413', 11)]

classes of webpages have been seen at least 300 times by the customers

In [19]:
%%sql 
SELECT webpageviewed, COUNT(*) AS views 
FROM adclicks 
GROUP BY webpageviewed 
HAVING views >= 300;

 * sqlite:///small_data/customers.sqlite
Done.


[('news', 412), ('social', 517)]

What adid/webpageviewed combinations have between 10 and 20 successful clicks (clicked value equal to 1)? (Hint: Use SELECT together with WHERE clicked=1, group the data appropriately, and add a HAVING clause.)

In [20]:
%%sql 
SELECT adid, webpageviewed 
FROM adclicks 
GROUP BY adid, webpageviewed
HAVING clicked = 1;

 * sqlite:///small_data/customers.sqlite
Done.


[(5, 'search'), (6, 'search')]

In [21]:
%%sql 
SELECT * 
FROM customers, adclicks           
WHERE customers.userid = adclicks.userid 
LIMIT 5;

 * sqlite:///small_data/customers.sqlite
Done.


[(64, 'Damien Cruise', '06535', 49, '2019-05-04 15:00:39', 9, 0, 'ecommerce', 64),
 (48, 'Gary Reynolds', '11731', 48, '2019-12-13 01:29:48', 8, 0, 'news', 48),
 (22, 'Beth Reynolds', '11715', 29, '2019-12-05 04:19:10', 10, 0, 'social', 22),
 (30, 'Charles Baker', '11731', -1, '2019-10-13 14:08:22', 6, 0, 'ecommerce', 30),
 (97, 'William Gunderson', '11715', 29, '2019-08-14 15:23:00', 2, 0, 'news', 97)]

In [22]:
%%sql
SELECT customers.userid, name, time, adid, clicked, webpageviewed 
FROM customers, adclicks
WHERE customers.userid = adclicks.userid 
LIMIT 5;

 * sqlite:///small_data/customers.sqlite
Done.


[(64, 'Damien Cruise', '2019-05-04 15:00:39', 9, 0, 'ecommerce'),
 (48, 'Gary Reynolds', '2019-12-13 01:29:48', 8, 0, 'news'),
 (22, 'Beth Reynolds', '2019-12-05 04:19:10', 10, 0, 'social'),
 (30, 'Charles Baker', '2019-10-13 14:08:22', 6, 0, 'ecommerce'),
 (97, 'William Gunderson', '2019-08-14 15:23:00', 2, 0, 'news')]

In [23]:
%%sql 
SELECT customers.userid, name, time, clicked, webpageviewed 
FROM customers, adclicks
WHERE customers.userid = adclicks.userid AND webpageviewed = 'news'
LIMIT 5;

 * sqlite:///small_data/customers.sqlite
Done.


[(48, 'Gary Reynolds', '2019-12-13 01:29:48', 0, 'news'),
 (97, 'William Gunderson', '2019-08-14 15:23:00', 0, 'news'),
 (44, 'George Baker', '2019-02-23 14:58:16', 0, 'news'),
 (89, 'Valerie Wallace', '2019-07-14 16:29:38', 0, 'news'),
 (84, 'Ann Cooper', '2019-09-01 10:36:56', 0, 'news')]

In [24]:
%%sql
SELECT * 
FROM customers NATURAL JOIN adclicks 
LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[(64, 'Damien Cruise', '06535', 49, '2019-05-04 15:00:39', 9, 0, 'ecommerce'),
 (48, 'Gary Reynolds', '11731', 48, '2019-12-13 01:29:48', 8, 0, 'news'),
 (22, 'Beth Reynolds', '11715', 29, '2019-12-05 04:19:10', 10, 0, 'social'),
 (30, 'Charles Baker', '11731', -1, '2019-10-13 14:08:22', 6, 0, 'ecommerce'),
 (97, 'William Gunderson', '11715', 29, '2019-08-14 15:23:00', 2, 0, 'news'),
 (44, 'George Baker', '11386', 42, '2019-02-23 14:58:16', 8, 0, 'news'),
 (84, 'Ann Cooper', '11386', 42, '2019-02-11 01:00:18', 8, 0, 'social'),
 (129, 'Katie Brand', '06794', 58, '2019-03-30 12:04:26', 6, 0, 'social'),
 (35, 'Cynthia Blofeld', None, 38, '2019-09-16 16:11:23', 8, 0, 'social'),
 (7, 'Katherine Diamond', '08605', 26, '2019-01-21 21:02:31', 10, 0, 'social')]

In [25]:
%%sql
SELECT customers.userid, name, time, adid, clicked, webpageviewed 
FROM customers NATURAL JOIN adclicks 
WHERE webpageviewed = 'news' AND customers.userid = 48 
LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[(48, 'Gary Reynolds', '2019-12-13 01:29:48', 8, 0, 'news'),
 (48, 'Gary Reynolds', '2019-05-16 09:49:37', 10, 1, 'news'),
 (48, 'Gary Reynolds', '2019-03-11 16:07:10', 9, 0, 'news'),
 (48, 'Gary Reynolds', '2019-06-02 20:29:02', 8, 0, 'news'),
 (48, 'Gary Reynolds', '2019-12-02 13:14:57', 5, 1, 'news')]

click-through rates for individual people for each category of webpage by selecting data and grouping it appropriately.

In [26]:
%%sql 
SELECT MIN(customers.userid), name, webpageviewed, 
    ROUND(SUM(clicked)*1.0/COUNT(clicked),3) AS click_rate
FROM customers NATURAL JOIN adclicks
WHERE customers.userid = 64 OR customers.userid = 48
GROUP BY name, webpageviewed;

 * sqlite:///small_data/customers.sqlite
Done.


[(64, 'Damien Cruise', 'ecommerce', 0.0),
 (64, 'Damien Cruise', 'news', 0.333),
 (64, 'Damien Cruise', 'search', 0.0),
 (64, 'Damien Cruise', 'social', 0.0),
 (48, 'Gary Reynolds', 'news', 0.4),
 (48, 'Gary Reynolds', 'search', 0.0),
 (48, 'Gary Reynolds', 'social', 0.222)]

It's unlikely that two customers will have the same "first view" time, but joining on the pair will ensure uniqueness in this case.

In [27]:
%%sql
WITH first_view_table AS 
   (SELECT userid, MIN(time) AS first_view 
   FROM customers JOIN adclicks USING(userid)
   GROUP BY userid)
SELECT *
FROM first_view_table JOIN adclicks ON first_view = time AND first_view_table.userid = adclicks.userid
LIMIT 5;

 * sqlite:///small_data/customers.sqlite
Done.


[(7, '2019-01-21 21:02:31', '2019-01-21 21:02:31', 10, 0, 'social', 7),
 (16, '2019-01-27 13:17:16', '2019-01-27 13:17:16', 7, 0, 'news', 16),
 (27, '2019-01-30 11:55:45', '2019-01-30 11:55:45', 6, 0, 'social', 27),
 (51, '2019-01-19 18:31:32', '2019-01-19 18:31:32', 6, 0, 'social', 51),
 (76, '2019-01-08 03:49:03', '2019-01-08 03:49:03', 6, 0, 'news', 76)]

count the number of "first views" for each type of website, or compute click-through rates for each type of website, etc.

In [28]:
%%sql

WITH first_view_table AS 
   (SELECT userid, MIN(time) AS first_view 
   FROM customers JOIN adclicks USING(userid)
   GROUP BY userid)
    
SELECT webpageviewed, COUNT(*) AS count_first_views
FROM first_view_table JOIN adclicks ON first_view = time AND first_view_table.userid = adclicks.userid
GROUP BY webpageviewed;

 * sqlite:///small_data/customers.sqlite
Done.


[('ecommerce', 9), ('news', 50), ('search', 17), ('social', 36)]

Computing click-through rates by website type

In [29]:
%%sql

WITH first_view_table AS 
   (SELECT userid, MIN(time) AS first_view 
   FROM customers JOIN adclicks USING(userid)
   GROUP BY userid)
    
SELECT webpageviewed, SUM(clicked) AS clicks, COUNT(*) AS views, 
   ROUND(AVG(clicked),3) AS first_click_through_rate
FROM first_view_table JOIN adclicks ON first_view = time AND first_view_table.userid = adclicks.userid
GROUP BY webpageviewed;

 * sqlite:///small_data/customers.sqlite
Done.


[('ecommerce', 1, 9, 0.111),
 ('news', 4, 50, 0.08),
 ('search', 4, 17, 0.235),
 ('social', 1, 36, 0.028)]

Counting click-through rates by adid

In [30]:
%%sql

WITH first_view_table AS 
   (SELECT userid, MIN(time) AS first_view 
   FROM customers JOIN adclicks USING(userid)
   GROUP BY userid)
SELECT adid, SUM(clicked) AS clicks, COUNT(*) AS views, 
   ROUND(AVG(clicked),3) AS first_click_through_rate
FROM first_view_table JOIN adclicks ON first_view = time AND first_view_table.userid = adclicks.userid
GROUP BY adid;

 * sqlite:///small_data/customers.sqlite
Done.


[(1, 0, 6, 0.0),
 (2, 0, 6, 0.0),
 (4, 1, 13, 0.077),
 (5, 1, 3, 0.333),
 (6, 3, 20, 0.15),
 (7, 0, 11, 0.0),
 (8, 2, 18, 0.111),
 (9, 1, 16, 0.063),
 (10, 2, 19, 0.105)]

we are generating a categorical variable based on a customer's age. We generate a new age_bracket column in the resulting table by checking the value of age and returning a new value.

In [31]:
%%sql
SELECT userid AS uid, age,
    CASE
        WHEN  0 <= age AND age < 13 THEN 'young'
        WHEN 13 <= age AND age <= 19 THEN 'teens'
        WHEN 20 <= age AND age < 30 THEN '20s'
        WHEN 30 <= age AND age < 40 THEN '30s'
        WHEN 40 <= age AND age < 50 THEN '40s'
        WHEN 50 <= age AND age < 60 THEN '50s'
        WHEN 60 <= age AND age < 70 THEN '60s'
        WHEN 70 <= age THEN 'seniors'
        ELSE 'other'
    END AS age_bracket 
FROM customers
LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[(1, 54, '50s'),
 (2, 20, '20s'),
 (3, 26, '20s'),
 (5, None, 'other'),
 (7, 26, '20s'),
 (8, 40, '40s'),
 (10, 72, 'seniors'),
 (11, 5, 'young'),
 (12, 53, '50s'),
 (13, 19, 'teens')]

In [32]:
%%sql 
WITH brackets AS 
    (SELECT userid AS uid, 
    CASE
        WHEN  0 <= age AND age < 13 THEN 'young'
        WHEN 13 <= age AND age <= 19 THEN 'teens'
        WHEN 20 <= age AND age < 30 THEN '20s'
        WHEN 30 <= age AND age < 40 THEN '30s'
        WHEN 40 <= age AND age < 50 THEN '40s'
        WHEN 50 <= age AND age < 60 THEN '50s'
        WHEN 60 <= age AND age < 70 THEN '60s'
        WHEN 70 <= age THEN 'seniors'
        ELSE 'other'
    END AS age_bracket 
    FROM customers)
    
SELECT age_bracket, 
    ROUND(AVG(clicked),3) AS click_rate, 
    COUNT(clicked) AS number_of_ad_views
FROM brackets, adclicks 
WHERE adclicks.userid = uid
GROUP BY age_bracket 
ORDER BY age_bracket;

 * sqlite:///small_data/customers.sqlite
Done.


[('20s', 0.156, 270),
 ('30s', 0.103, 312),
 ('40s', 0.13, 270),
 ('50s', 0.086, 267),
 ('other', 0.104, 48),
 ('seniors', 0.0, 18),
 ('teens', 0.101, 109),
 ('young', 0.148, 27)]

find click-through rates for different age group/adid combinations.

In [34]:
%%sql 
WITH brackets AS 
    (SELECT userid AS uid, 
    CASE
        WHEN  0 <= age AND age < 13 THEN 'young'
        WHEN 13 <= age AND age <= 19 THEN 'teens'
        WHEN 20 <= age AND age < 30 THEN '20s'
        WHEN 30 <= age AND age < 40 THEN '30s'
        WHEN 40 <= age AND age < 50 THEN '40s'
        WHEN 50 <= age AND age < 60 THEN '50s'
        WHEN 60 <= age AND age < 70 THEN '60s'
        WHEN 70 <= age THEN 'seniors'
        ELSE 'other'
    END AS age_bracket 
    FROM customers)
SELECT adid, age_bracket, 
    ROUND(AVG(clicked),3) AS click_rate, 
    COUNT(clicked) AS number_of_ad_views
    
FROM brackets, adclicks 
WHERE adclicks.userid = uid
GROUP BY adid, age_bracket 
ORDER BY adid, age_bracket
LIMIT 10;

 * sqlite:///small_data/customers.sqlite
Done.


[(1, '20s', 0.077, 13),
 (1, '30s', 0.071, 14),
 (1, '40s', 0.1, 10),
 (1, '50s', 0.091, 11),
 (1, 'other', 0.0, 1),
 (1, 'seniors', 0.0, 2),
 (1, 'teens', 0.0, 5),
 (2, '20s', 0.077, 13),
 (2, '30s', 0.1, 20),
 (2, '40s', 0.143, 14)]

In [36]:
%%sql 
SELECT age/10 AS decades_old, 
    ROUND(AVG(clicked),3) AS click_rate, 
    COUNT(clicked) AS number_of_ad_views
    
FROM customers, adclicks 
WHERE customers.userid = adclicks.userid
GROUP BY decades_old 
ORDER BY decades_old;

 * sqlite:///small_data/customers.sqlite
Done.


[(None, 0.128, 39),
 (0, 0.111, 36),
 (1, 0.101, 109),
 (2, 0.156, 270),
 (3, 0.103, 312),
 (4, 0.13, 270),
 (5, 0.086, 267),
 (7, 0.0, 8),
 (8, 0.0, 10)]

SQL Alchemy

In [45]:
import sqlalchemy

In [46]:
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, String

# This creates a base we can inherit from
Base = declarative_base()

class Customers(Base):
    __tablename__ = "customers"
    
    # sqlalchemy uses these names as the column names
    userID = Column(Integer, primary_key=True)
    name = Column(String)
    zipcode = Column(String)
    age = Column(Integer)

In [47]:
engine = sqlalchemy.create_engine('sqlite:///small_data/customers.sqlite', echo=True)

session_gen = sqlalchemy.orm.sessionmaker(bind=engine)

In [48]:
session = session_gen()

In [49]:
%%sql

SELECT userid, name
FROM customers
WHERE userid < 40
ORDER BY userid DESC
LIMIT 7

 * sqlite:///small_data/customers.sqlite
Done.


[(39, 'Julia Powers'),
 (38, 'Mary Reynolds'),
 (37, 'Rob Frye'),
 (36, 'Chuck Finley'),
 (35, 'Cynthia Blofeld'),
 (34, 'Valerie Clarkson'),
 (33, 'William Brand')]

In [50]:
results = session.query(Customers.userID, Customers.name).filter(Customers.userID < 40)\
                  .order_by(sqlalchemy.desc(Customers.userID))[2:7]

for row in results:
    print(row)

2021-12-10 11:01:41,434 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2021-12-10 11:01:41,435 INFO sqlalchemy.engine.base.Engine ()
2021-12-10 11:01:41,436 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2021-12-10 11:01:41,436 INFO sqlalchemy.engine.base.Engine ()
2021-12-10 11:01:41,437 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2021-12-10 11:01:41,437 INFO sqlalchemy.engine.base.Engine SELECT customers."userID" AS "customers_userID", customers.name AS customers_name 
FROM customers 
WHERE customers."userID" < ? ORDER BY customers."userID" DESC
 LIMIT ? OFFSET ?
2021-12-10 11:01:41,438 INFO sqlalchemy.engine.base.Engine (40, 5, 2)
(37, 'Rob Frye')
(36, 'Chuck Finley')
(35, 'Cynthia Blofeld')
(34, 'Valerie Clarkson')
(33, 'William Brand')


In [51]:
session.query(Customers.name, Customers.zipcode, Customers.age)\
       .filter(sqlalchemy.or_(Customers.zipcode.like('06%'), Customers.zipcode == None))\
       .filter(Customers.age.in_([19, 20, 21, 22, 23])).all()

2021-12-10 11:01:57,785 INFO sqlalchemy.engine.base.Engine SELECT customers.name AS customers_name, customers.zipcode AS customers_zipcode, customers.age AS customers_age 
FROM customers 
WHERE (customers.zipcode LIKE ? OR customers.zipcode IS NULL) AND customers.age IN (?, ?, ?, ?, ?)
2021-12-10 11:01:57,786 INFO sqlalchemy.engine.base.Engine ('06%', 19, 20, 21, 22, 23)


[('Gabriel Thomasson', None, 19),
 ('Violet Green', '06794', 20),
 ('Valerie Frye', None, 19),
 ('Mary Reynolds', '06794', 23),
 ('Irina Johansen', '06410', 21),
 ('Marie Brand', '06794', 19)]

In [52]:
from sqlalchemy.sql import func

underage = session.query(Customers).filter(Customers.age < 21).count()
print("There are {} people who are under 21".format(underage))

session.query(Customers.zipcode, func.count('*')).filter(Customers.zipcode.like('06%'))\
       .group_by(Customers.zipcode).all()

2021-12-10 11:02:12,890 INFO sqlalchemy.engine.base.Engine SELECT count(*) AS count_1 
FROM (SELECT customers."userID" AS "customers_userID", customers.name AS customers_name, customers.zipcode AS customers_zipcode, customers.age AS customers_age 
FROM customers 
WHERE customers.age < ?) AS anon_1
2021-12-10 11:02:12,891 INFO sqlalchemy.engine.base.Engine (21,)
There are 16 people who are under 21
2021-12-10 11:02:12,892 INFO sqlalchemy.engine.base.Engine SELECT customers.zipcode AS customers_zipcode, count(?) AS count_1 
FROM customers 
WHERE customers.zipcode LIKE ? GROUP BY customers.zipcode
2021-12-10 11:02:12,893 INFO sqlalchemy.engine.base.Engine ('*', '06%')


[('06245', 2),
 ('06267', 4),
 ('06403', 4),
 ('06410', 2),
 ('06535', 6),
 ('06794', 8)]

In [53]:
case_part = sqlalchemy.case([(Customers.age < 20, "young"), \
                               (sqlalchemy.and_(Customers.age >= 20, Customers.age < 30), "20s")], \
                            else_="older").label("age_group")

session.query(Customers.name, Customers.age, case_part)[:10]

2021-12-10 11:02:21,170 INFO sqlalchemy.engine.base.Engine SELECT customers.name AS customers_name, customers.age AS customers_age, CASE WHEN (customers.age < ?) THEN ? WHEN (customers.age >= ? AND customers.age < ?) THEN ? ELSE ? END AS age_group 
FROM customers
 LIMIT ? OFFSET ?
2021-12-10 11:02:21,171 INFO sqlalchemy.engine.base.Engine (20, 'young', 20, 30, '20s', 'older', 10, 0)


[('Katie Baker', 54, 'older'),
 ('Mary Clarkson', 20, '20s'),
 ('Michelle Reynolds', 26, '20s'),
 ("Valerie O'Brian", None, 'older'),
 ('Katherine Diamond', 26, '20s'),
 ('Imran Brown', 40, 'older'),
 ('Francis Baker', 72, 'older'),
 ('Rob Smith', 5, 'young'),
 ('Clare Johnson', 53, 'older'),
 ('Gabriel Thomasson', 19, 'young')]